### PR TITLE
feat: add WASM transformer

### PR DESCRIPTION
**Pull request by Soumil-07**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -102,7 +102,7 @@ build: get $(BINDIR)/$(BINNAME) ## Build go code
 	@printf "\033[32m-------------------------------------\n BUILD SUCCESS\n-------------------------------------\033[0m\n"
 
 $(BINDIR)/$(BINNAME): $(SRC) $(ASSETS) $(WEB_ASSETS)
-	CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o $(BINDIR)/$(BINNAME) .
+	CGO_ENABLED=1 go build -ldflags '$(LDFLAGS)' -o $(BINDIR)/$(BINNAME) .
```
Comment (HarikrishnanBalagopal) [inline Makefile:105]: Can we keep `CGO_ENABLED=0` ?

File: go.mod [hunk 2]
```diff
@@ -184,6 +184,7 @@ require (
 	github.com/robfig/cron v1.2.0 // indirect
 	github.com/russross/blackfriday v1.6.0 // indirect
 	github.com/russross/blackfriday/v2 v2.1.0 // indirect
+	github.com/second-state/WasmEdge-go v0.13.0 // indirect
```
Comment (HarikrishnanBalagopal) [inline go.mod:187]: Is this indirect or direct?
Try using `go mod tidy`

File: transformer/external/wasmtransformer.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	EnvList    []core.EnvVar `yaml:"env,omitempty"`
+}
+
+// Init Initializes the transformer
+func (t *WASM) Init(tc transformertypes.Transformer, env *environment.Environment) (err error) {
+	t.Config = tc
+	t.Env = env
+	t.WASMConfig = &WASMYamlConfig{}
+	if err := common.GetObjFromInterface(t.Config.Spec.Config, t.WASMConfig); err != nil {
+		return fmt.Errorf("unable to load config for Transformer %+v into %T . Error: %q", t.Config.Spec.Config, t.WASMConfig, err)
```
Comment (HarikrishnanBalagopal) [inline transformer/external/wasmtransformer.go:61]: `Error: %q`
use `%w` to wrap the error

File: transformer/external/wasmtransformer.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	mem := mod.FindMemory("memory")
+	memData, err := mem.GetData(uint(dirPointer), uint(len(dir)+1))
+	if err != nil {
+		return nil, err
+	}
+	copy(memData, dir)
+	memData[len(dir)] = 0
+	directoryDetectOutput, dderr := vm.Execute("directoryDetect", dirPointer)
+	if dderr != nil {
+		err = fmt.Errorf("failed to execute directoryDetect in the wasm module. Error : %s", dderr.Error())
```
Comment (HarikrishnanBalagopal) [inline transformer/external/wasmtransformer.go:102]: Use `%w` to wrap the error.
`Error : %w", dderr)`

File: transformer/external/wasmtransformer.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	}
+	resultPointer := binary.LittleEndian.Uint32(memData[:4])
+	resultLength := binary.LittleEndian.Uint32(memData[4:])
+	memData, err = mem.GetData(uint(resultPointer), uint(resultLength))
+	if err != nil {
+		return nil, nil, err
+	}
+	logrus.Debug(string(memData))
+	var output transformertypes.TransformOutput
+	json.Unmarshal(memData, &output)
```
Comment (HarikrishnanBalagopal) [inline transformer/external/wasmtransformer.go:206]: Check the error returned by `json.Unmarshal`

**Review by HarikrishnanBalagopal — CHANGES_REQUESTED** [review]
Mostly fine. Just some minor changes.

File: Makefile (hunk 1 above)
Comment (Soumil-07) [reply Makefile:105]: I couldn't build WASMEdge with `CGO_ENABLED=1`: 

```
github.com/second-state/WasmEdge-go/wasmedge: build constraints exclude all Go files in /Users/<redacted>/go/pkg/mod/github.com/second-state/!wasm!edge-go@v0.13.0/wasmedge```

File: go.mod (hunk 2 above)
Comment (Soumil-07) [reply go.mod]: Resolved in d87520b

File: transformer/external/wasmtransformer.go (hunk 4 above)
Comment (Soumil-07) [reply transformer/external/wasmtransformer.go]: Resolved in d87520b

File: transformer/external/wasmtransformer.go (hunk 5 above)
Comment (Soumil-07) [reply transformer/external/wasmtransformer.go]: Resolved in d87520b

File: transformer/external/wasmtransformer.go (hunk 3 above)
Comment (Soumil-07) [reply transformer/external/wasmtransformer.go]: Resolved in d87520b

File: Makefile (hunk 1 above)
Comment (HarikrishnanBalagopal) [reply Makefile:105]: ok wasmedge-go seems to be a binding on the C++ wasmedge project

**Comment by HarikrishnanBalagopal** [comment]
Possibly we can move to https://github.com/tetratelabs/wazero which is pure Go with no dependency.

**Comment by HarikrishnanBalagopal** [comment]
With WASI preview 2 being officially released https://github.com/WebAssembly/WASI/tree/main/preview2 it brings the WASM component model.

Possibly we can move to the WASM component model for marshalling and unmarshalling https://component-model.bytecodealliance.org/design/why-component-model.html

**Comment by HarikrishnanBalagopal** [comment]
Related https://github.com/konveyor/move2kube/pull/1144

**Comment by HarikrishnanBalagopal** [comment]
Implemented in https://github.com/konveyor/move2kube/pull/1172

**Pull request closed without merging**